### ISSUE TITLE
create local genesis devnets

### ISSUE THREAD

**Issue by dzmitry-lahoda** [informalsystems/cosmos.nix#168]
there are some commont scripts place like

https://github.com/cosmology-tech/starship/blob/main/charts/devnet/scripts/neutron/create-neutron-genesis.sh

or custom

https://github.com/osmosis-labs/osmosis/blob/main/tests/localosmosis/scripts/setup.sh

these scripts allow to setup local devnet

so cosmos.nix could use these to provide default devnets on top of just binaries

**Comment by dzmitry-lahoda** [comment]
nix does not need docker or helm, but nix can reuse starship bash and be like starship

**Comment by JonathanLorimer** [comment]
I also looked into creating nixos modules to facilitate nixos tests (which support arbitrary number of nodes via qemu). Unfortunately spinning up chains is extremely imperative and stateful and was hard for me to get right. I think there are a lot of solutions that could work here.

**Comment by dzmitry-lahoda** [comment]
yeah. port binding is one of issues which hard to handle. so in theory can do most common/representative nodes so(cosmos hub, cosmwasm node). 
so reusing existing sh files worked well. and starship genesises work well enough. but need more thinking. 


>  nixos tests (which support arbitrary number of nodes via qemu).

nice, where can I read about this? never saw such setup

**Comment by JonathanLorimer** [comment]
@dzmitry-lahoda the documentation is really awful, but here is a small piece https://nix.dev/tutorials/nixos/integration-testing-using-virtual-machines.html

You can also look some of the experimental work I did [here](https://github.com/informalsystems/cosmos.nix/blob/master/modules/tests/hermes-test.nix) and [here](https://github.com/informalsystems/cosmos.nix/blob/master/modules/tests/gaia-test.nix). Not shown in the git tree is that I manually generated the data dirs for the chains and simply referenced them in the tests. Ideally initializing the chains would be reproducible and declarative as well.

**Comment by dzmitry-lahoda** [comment]
> I think there are a lot of solutions that could work here.

yeah. seems direct reuse of bash in bad idea in general eventually. but assuming `startship` catches up, it is best one, just follow starship.  but as of now not so many published startship modules. so it may be deadend.

as of nix, it can git reuse bash from chains from start.

and later it can have nixos modules with default and override accounts and bash sub scripts to init account amounts.
as of now I maintain 3 chains in simpler way, will have to do afaik +2 more. so total 5 cosmos chains.
my current approach is stupid - copy bash into my nix, use ${....} to plug shared parameters.
run post gen pos run scripts to setup IBCs

**Comment by JonathanLorimer** [comment]
I think there needs to be more planning around this, but its something we should support. Would you be willing to either sketch out the approach in pseudo code / a comment, or propose a simple PR with a proof of concept?

**Comment by dzmitry-lahoda** [comment]
I guess alignment is first.

I will described how I ended up in https://github.com/ComposableFi/composable/blob/main/flake/process-compose.nix .

Before I ended up with process-compose solution, I considered(and tried) docker/podman/systemd/launchd/k8s and several others process managers. I have experience in k8s, but not local.

There are 2 faces - end user functional area(QA/xdapps deps/FE) and developer area(nix developer).

**Comment by dzmitry-lahoda** [comment]
I think end user wants `lightweight portable any CPU rootless zero config dashboarded way to run set of process`.

So:
docker - needs root to setup and some config, not lightweights until build OCI image which is not container
k8s - how same config portable accross all machines

both above are fat (need big image), so alfa hacks not to build image like this nix-snapshotter or some articles about OCI structure.

so not really ready or cool.

podman - config, without root possible, but limited, not slick/documented/compatibe as docker

systemd/launchd - seems may be, as soon as some `nixd` project will unify both under same module, so it is blocker. i am nix tooling user - not writer. but if such exists - i would look at it very sure. problem - not all linuxes are systemd.

pure linux sucks as process manager, there are several attemps to write own (all so so tbh), and some like parity wrote their own (zombienet). the have DSL which allows to emit podman/k8s/native (native basically their own written process manager - it sucks tbh - no dashboard).

so i stayed with process-manager-flake - risk it is only 300 starts, not 3000k, so dev is reponsive.

**Comment by dzmitry-lahoda** [comment]
Another topic for nix devs, is how to stucture configs and share/reuse genesis overrides/generators.

Structure codes - process-compose-flake and terranix has nice examples of modules usage out of nixos, and also https://discourse.nixos.org/t/best-resources-for-learning-about-the-nixos-module-system/1177

Seems like each chain to be module for dynamic composition via eval (my current nix is static). But need to design modules first.

Other question how to share scripts?

Reuse from repo of chain team? From starship? Wrap TS/Python they have for cooking as nix commands? Write own bash? Generated from scrach in pure nix? Cosmos SDK genesis builder?

I found using `bash(and linux cli tools) and anything chain team` has is best approach. Do not write TS, wait TS/GO/Python tools emerge works for me.

So need design for modules and pick 2/3 chains to start first (Cosmos Hub/CosmWasm simulator seems may be good to go for devs).

And also realyers not dicussed - so bash and tools to for me.

At least this apprach supports ether/polkadot/cosmsos and soon(sure sure) solana ibc.

**Comment by dzmitry-lahoda** [comment]
Also I think these people are ready https://developers.openibc.com/docs/category/ibctl-cli-tool ? I think they are alfa and not flexible.

Same as osmosis beaker.

So really nix + modules + bash(dasel/jq/base64/tar+each teach tooling) + process compose is like way to processed.

Modules need desing

**Comment by dzmitry-lahoda** [comment]
@JonathanLorimer  TLRD;

First need to align on process manager, I am for process-compose; OR for OCI with only link to /nix/store, but not all closures copied, whatever runner - k8s, docker, podman; OR systemd/launchd. 

OCI as it is built

**Comment by dzmitry-lahoda** [comment]
here is module model for process-compose https://github.com/juspay/services-flake

so not sure if it is good one.

how to express graph dependencies, for example that relayer depends on 2 chains?

**Comment by dzmitry-lahoda** [comment]
as i need this now and multi chain (not only cosmos) and need from shell from devnet to mainnet, i go for now:

- https://github.com/ComposableFi/networks
- process-compose

please consider moving cosmos.nix to cosmos org to allow improved collaboration on devnets if needed.

**Comment by dzmitry-lahoda** [comment]
as i see set of configs can be

```sh
BINARY=${BINARY:-neutrond}
BASE_DIR=./data
CHAINID=${CHAINID:-test-1}
FEEDENOM=STAKEDENOM
STAKEDENOM=${STAKEDENOM:-untrn}
CHAIN_DIR="$BASE_DIR/$CHAINID"

P2PPORT=${P2PPORT:-26656}
RPCPORT=${RPCPORT:-26657}
RESTPORT=${RESTPORT:-1317}
ROSETTA=${ROSETTA:-8081}
``
kind of cosmos-sdk subset.

**Comment by dzmitry-lahoda** [comment]
other things - jq/sed is bearshish. sed may overwrite good value. so seems dasel is most structured and universal for rewrite of toml/json/yaml.
